### ISSUE TITLE
bug(models): failed image prefill returns Ok(steps) — gemma3/gemma4 swallow the cause AND can fake a success

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#242]
Surfaced by the #235 review, deliberately left out of that PR's scope. **Same swallow class as #235, but with a sharper edge.**

## Problem
`crates/rmlx-models/src/gemma3/generate.rs:276` and `crates/rmlx-models/src/gemma4/generate/mod.rs:467` do `return Ok(steps)` when **image prefill fails**, discarding the real error.

#235 fixed the decode loops (`warn!` + `break` → `Ok(short)` → server reported `finish_reason:"length"`). Its companion instance — the *chunked text prefill* swallow in bitnet/laguna/qwen2 — is being fixed alongside it; there `steps` is always empty, so the engine backstop (`arch_generator.rs:1411`: `Ok(steps) if steps.is_empty()` → `Error::Other("generation produced zero tokens")`) at least keeps a fake success off the wire (it still erases the cause, and misclassifies as `Migratable` → 2 doomed replays).

**Here it is worse:** on the vision path `steps` is **not always empty**, so the backstop does not fire. A failed image prefill can return a *successful-looking short generation* — the exact `finish_reason:"length"` masking #235 exists to kill, on the multimodal surface.

## Why it matters
This is the bug that let #217 ship a live breakage and let `iso3` silently corrupt output for a release (see #233): a failure that reports as a success defeats every automated gate we have. The vision path has no equivalent of the canary, so it would be even less likely to surface.

## Fix
Propagate, matching #235's decode/prefill treatment:
```rust
Err(e) => {
    tracing::error!(error = %e, "image prefill failed, aborting generation");
    return Err(e);
}
```
Model-agnostic; applies to both arches identically.

## Test
- Force an image-prefill failure on a vision-capable model (gemma-4-12b-unified or a gemma3 vision snapshot) → the response must NOT be indistinguishable from a clean short generation; the real cause must reach the caller.
- A normal image request still succeeds, coherent at temp=0 (≥1 vision arch, streaming + non-streaming).
- If #235's `make check-no-decode-swallow` grep gate lands, extend it to catch this `return Ok(steps)`-after-error shape too.

Related: #235 (same class), #217 / #233 (what this masking cost us).

**Comment by Pushkinist** [comment]
**Already fixed by #243 (merged #253, `dee2dad`).** Verified in source on current `main`:

- `crates/rmlx-models/src/gemma4/generate/mod.rs:459-486` — image-prefill forward now captures `first_err` on failure, runs the mandatory `exit_prefill` sweep over **all** caches, then `if let Some(e) = first_err { return Err(e); }`. Only reaches the logits on clean success. No `return Ok(steps)` swallow remains.
- `crates/rmlx-models/src/gemma3/generate.rs:275-296` — identical shape.

This is exactly the propagate-not-swallow fix #242 asked for; #243's deferred-`first_err` design applied to the vision path closes both the swallow AND the strand (exit_prefill runs even on failure). The remaining `return Ok(steps)` lines in those files are legitimate decode-completion paths, not the image-prefill swallow.

Closing as fixed. One residual, folded into a later gate-hardening pass (not reopening this): extend `check-no-decode-swallow` to also catch the `return Ok(steps)`-after-error shape on the vision path, so this can't silently regress.
